domain/personallearningplan/src/main/kotlin/uk/gov/justice/digital/hmpps/domain/personallearningplan/service/GoalService.kt

```diff
@@ -84,6 +99,71 @@ class GoalService(
       }
   }
 
+  /**
+   * Archives a [Goal], identified by its `prisonNumber` and `goalReference`, from the specified [ArchiveGoalDto].
+   */
+  fun archiveGoal(prisonNumber: String, archiveGoalDto: ArchiveGoalDto): Either<ArchiveGoalProblem, Goal> {
+    val goalReference = archiveGoalDto.reference
+    log.info { "Archiving Goal with reference [$goalReference] for prisoner [$prisonNumber] because [${archiveGoalDto.reason}] with description [${archiveGoalDto.reasonOther ?: ""}]" }
+
+    return goalPersistenceAdapter.getGoal(prisonNumber, goalReference)
```

Comment:
Oooohhhh, how does `goalPersistenceAdapter.getGoal` return something that has a `.orProblem` method?
```
return goalPersistenceAdapter.getGoal(prisonNumber, goalReference)
      .orProblem(GoalToBeArchivedCouldNotBeFound(prisonNumber, goalReference))
      ....
```

As far as I can tell in this PR the `getGoal` method in the persistence adapter has not been changed and returns a nullable `Goal` ?
```
  /**
   * Returns a [Goal] identified by its `prisonNumber` and `goalReference` if found, otherwise `null`.
   */
  fun getGoal(prisonNumber: String, goalReference: UUID): Goal?
```

Has the inclusion of the `arrows` library essentially added an extension function or similar on every type? 😬

Reply:
Doh! scrap that!, I've just seen where you add the extension function towards the end of this class
Panic over! 🤣

Reply:
Not quite panic over .... the ext function you added below is this:
```
private fun <T> Goal?.orProblem(problem: T): Either<T, Goal> = this?.right() ?: problem.left()
```

`this` in  `this?.right()` is the `Goal`; therefore something somewhere has added a `right` function to the `Goal` (and presumably also a `left` function)
Clicking into the `right` function it takes me to the `arrows` source code:
```
public fun <A> A.right(): arrow.core.Either<kotlin.Nothing, A> { /* compiled code */ }
```
It seems that every class now has a `left` and `right` function (and all the others that presumably `arrows` has augmented into them) 🤔 . EG. using typeahead I can see that the `UUID` class now has these functions:

![Screenshot 2024-06-25 at 08 28 26](https://github.com/ministryofjustice/hmpps-education-and-work-plan-api/assets/94835226/76c60632-2570-4c09-8498-030d42a011fe)

Reply:
Yeah left() and right() are just extension functions provided by arrow that create Either.Left or Either.Right wrappers around your type. They are much more concise for things like that orProblem method I created. It's a more functional style so appreciate it might take some getting used to.